internal/venice-common/src/main/java/com/linkedin/venice/pubsub/manager/TopicMetadataFetcher.java

```diff
@@ -0,0 +1,673 @@
+      throw new IllegalArgumentException("Invalid partition number: " + pubSubTopicPartition.getPartitionNumber());
+    }
+    if (!containsTopicCached(pubSubTopicPartition.getPubSubTopic())) {
+      throw new PubSubTopicDoesNotExistException(
+          "Either topic: " + pubSubTopicPartition.getPubSubTopic() + " does not exist or partition: "
+              + pubSubTopicPartition.getPartitionNumber() + " is invalid");
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+    latestOffsetCache.clear();
+    earliestOffsetCache.clear();
+    lastProducerTimestampCache.clear();
+    topicExistenceCache.clear();
+
+    threadPoolExecutor.shutdown();
+    try {
+      if (threadPoolExecutor.awaitTermination(5, TimeUnit.MILLISECONDS)) {
+        threadPoolExecutor.shutdownNow();
+      }
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+    }
+
+    for (Closeable closeable: closeables) {
+      Utils.closeQuietlyWithErrorLogged(closeable);
+    }
+  }
+
+  // API: Check if a topic exists in the pubsub cluster
+  boolean containsTopic(PubSubTopic topic) {
+    long startTime = System.currentTimeMillis();
+    boolean containsTopic = pubSubAdminAdapter.containsTopic(topic);
+    TopicManagerStats.recordLatency(
+        topicManagerStats,
+        TopicManagerStats.OCCURRENCE_LATENCY_SENSOR_TYPE.CONTAINS_TOPIC,
+        System.currentTimeMillis() - startTime);
+    return containsTopic;
+  }
+
+  CompletableFuture<Boolean> containsTopicAsync(PubSubTopic topic) {
+    return CompletableFuture.supplyAsync(() -> containsTopic(topic), threadPoolExecutor);
+  }
+
+  boolean containsTopicCached(PubSubTopic pubSubTopic) {
+    long now = System.nanoTime();
+    ValueAndExpiryTime<Boolean> cachedValue = topicExistenceCache.computeIfAbsent(
+        pubSubTopic,
+        k -> new ValueAndExpiryTime<>(containsTopic(pubSubTopic), now + cachedEntryTtlInNs));
+    if (cachedValue.getExpiryTimeNs() <= now && cachedValue.tryAcquireUpdateLock()) {
+      CompletableFuture<Boolean> containsTopicCompletableFuture = containsTopicAsync(pubSubTopic);
+      containsTopicCompletableFuture.whenComplete((containsTopic, throwable) -> {
+        if (throwable != null) {
+          topicExistenceCache.remove(pubSubTopic);
+        } else {
+          topicExistenceCache
+              .put(pubSubTopic, new ValueAndExpiryTime<>(containsTopic, System.nanoTime() + cachedEntryTtlInNs));
+        }
+      });
+    }
+    return cachedValue.getValue();
+  }
+
+  // API: Get the latest offsets for all partitions of a topic
+  /**
+   * Get the latest offsets for all partitions of a topic. This is a blocking call.
+   * @param topic topic to get latest offsets for
+   * @return a map of partition id to latest offset. If the topic does not exist, an empty map is returned.
+   */
+  Int2LongMap getTopicLatestOffsets(PubSubTopic topic) {
+    PubSubConsumerAdapter pubSubConsumerAdapter = acquireConsumer();
```

Comment:
`acquireConsumer()` works on a `BlockingQueue` and is expected to return consumer client or wait until one becomes available.  It is expected to throw only `InterruptedException`, in which it doesn't remove consumer from the blocking queue.  So IMO having it in `try` doesn't add any value.  

We do need to make sure that `releaseConsumer` is called every time control reaches a `try` block`.

We could consider wrapping `BlockingQueue<PubSubConsumerAdapter>` in `AutoClosable` `PubSubConsumerResource` so that we can use it in `try-with-resource`. The cost will be an additional wrapper object per function call.